xarray_beam/_src/pangeo.py

```diff
@@ -0,0 +1,150 @@
+"""IO with Pangeo-Forge."""
+from typing import (
+  Dict,
+  Iterator,
+  Optional,
+  Mapping,
+  Tuple,
+  cast,
+)
+
+import apache_beam as beam
+import xarray
+from apache_beam.io.filesystems import FileSystems
+
+from xarray_beam._src import core
+
+
+def _zero_dimensions(dataset: xarray.Dataset) -> Mapping[str, int]:
+  return {dim: 0 for dim in dataset.dims.keys()}
+
+
+def _expand_dimensions_by_key(
+    dataset: xarray.Dataset,
+    key: core.Key,
+    index: Tuple[int, ...],
+    pattern: 'FilePattern'
+) -> xarray.Dataset:
+  """Expand the dimensions of the `Dataset` by offsets found in the `Key`."""
+  combine_dims_by_name = {
+    combine_dim.name: (i, combine_dim)
+    for i, combine_dim in enumerate(pattern.combine_dims)
+  }
+
+  if not combine_dims_by_name:
+    return dataset
+
+  for dim_key in key.offsets.keys():
+    # skip expanding dimensions if they already exist
+    if dim_key in dataset.dims:
+      continue
+
+    dim_idx, combine_dim = combine_dims_by_name.get(dim_key, (-1, None))
+    if dim_idx == -1:
+      raise ValueError(
+        f"could not find CombineDim named {dim_key!r} in pattern {pattern!r}."
+      )
+
+    dim_val = combine_dim.keys[index[dim_idx]]
+    dataset = dataset.expand_dims(**{dim_key: [dim_val]})
+
+  return dataset
+
+
+class FilePatternToChunks(beam.PTransform):
+  """Open data described by a Pangeo-Forge `FilePattern` into keyed chunks."""
+
+  from pangeo_forge_recipes.patterns import FilePattern
+
+  def __init__(
+      self,
+      pattern: 'FilePattern',
+      sub_chunks: Optional[Mapping[str, int]] = None,
+      xarray_open_kwargs: Optional[Dict] = None
+  ):
+    """Initialize FilePatternToChunks.
+
+    TODO(#29): Currently, `MergeDim`s are not supported.
+
+    Args:
+      pattern: a `FilePattern` describing a dataset.
+      sub_chunks: split each open dataset into smaller chunks. If not set, each
+        chunk will open the full dataset.
+      xarray_open_kwargs: keyword arguments to pass to `xarray.open_dataset()`.
+    """
+    self.pattern = pattern
+    self.sub_chunks = sub_chunks or -1
+    self.xarray_open_kwargs = xarray_open_kwargs or {}
+
+  def _prechunk(self) -> Iterator[Tuple[core.Key, Tuple[int, ...]]]:
+    """Converts `FilePattern` items into keyed indexes."""
+    input_chunks = {k: v or 1 for k, v in self.pattern.nitems_per_input.items()}
+    dim_sizes = {
+      k: v or self.pattern.dims[k]
+      for k, v, in self.pattern.concat_sequence_lens.items()
+    }
+    chunks = core.normalize_expanded_chunks(input_chunks, dim_sizes)
+    for key, (index, _) in zip(core.iter_chunk_keys(chunks),
+                               self.pattern.items()):
+      yield key, index
+
+  def _open_chunks(
+      self,
+      key: core.Key,
+      index: Tuple[int, ...]
+  ) -> Iterator[Tuple[core.Key, xarray.Dataset]]:
+    """Open datasets into chunks with XArray."""
+    path = self.pattern[index]
+    with FileSystems().open(path) as file:
```

Comment:
Do Beam's filesystems really all work out of the box with Xarray? If so, that's awesome!

Can you verify that it works with both netCDF3 and netCDF4 files? These would be using different underlying storage backends (scipy vs h5netcdf).

To be honest, I'm a little skeptical that this will work well. I suspect we'll end up up needing to copy temporary files to local disk (but I'd love to be proven wrong!)

Reply:
Let me experiment and see how this works. In my tests in the previous iteration of this change, this worked well with GCS's IO objects.

Reply:
It's a bit unclear to me how this would not work in a Beam pipeline (or, what needs to be done to get this win). Can you explain a bit more? 

Is this a correct understanding: With the change you're referring to, we could pickle the XArray open command (with the file-like object) as PCollections, which would allow us to split the open across workers?

Reply:
> Is this a correct understanding: With the change you're referring to, we could pickle the XArray open command (with the file-like object) as PCollections, which would allow us to split the open across workers?

With this change, we could pickle lazy `xarray.Dataset` objects corresponding to open netCDF files and pass them between stages in in a Beam pipeline.

Some data would still need to get loaded on worker on which `xarray.open_dataset()` is called, but this could be much less data than the entire file (e.g., only the "metadata" part of the file). The bulk of the loading work could be split across multiple workers, which could be quite useful for processing large (GB+) netCDF files.